Refactor, second episode: move to PSBTs

darosior opened the pull request:
This is based on #10 (as pretty much everything now..) and implements the usage of PSBT in our `RevaultTransaction` management, from the creation to the satisfaction.

This unlocked a few side improvements, such as:
- Get rid of the RevaultSatisfier struct
- Libbitcoinconsensus verification is back!
- Even more typesafety (eg for TxOuts and script descriptors)

The methods of the `RevaultTransaction` traits have been modified with [BIP174 roles in mind](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#responsibilities), which will likely guide us toward less footguns downstream.

darosior commented:
Rebased after #10 rebase.

darosior commented:
Rebased after #10 merge :tada:

darosior commented on src/prevouts.rs:
```diff
@@ -1,33 +1,43 @@
+        #[derive(Debug, Clone)]
+        pub struct $struct_name<T: RevaultTxOut> {
+            outpoint: OutPoint,
+            prev_txout: T,
```
Makes much more sense this way:
```patch
diff --git a/src/transactions.rs b/src/transactions.rs
index b9224d3..3d65127 100644
--- a/src/transactions.rs
+++ b/src/transactions.rs
@@ -377,7 +377,7 @@ impl UnvaultTransaction {
     /// addition to the unvault one.
     /// PSBT Creator and Updater.
     pub fn new(
-        vault_input: VaultTxIn<VaultTxOut>,
+        vault_input: VaultTxIn,
         unvault_txout: UnvaultTxOut,
         cpfp_txout: CpfpTxOut,
         lock_time: u32,
@@ -399,8 +399,8 @@ impl CancelTransaction {
     /// may have a fee-bumping input.
     /// PSBT Creator and Updater.
     pub fn new(
-        unvault_input: UnvaultTxIn<UnvaultTxOut>,
-        feebump_input: Option<FeeBumpTxIn<FeeBumpTxOut>>,
+        unvault_input: UnvaultTxIn,
+        feebump_input: Option<FeeBumpTxIn>,
         vault_txout: VaultTxOut,
         lock_time: u32,
     ) -> CancelTransaction {
@@ -432,8 +432,8 @@ impl EmergencyTransaction {
     /// Script. It may also spend an additional output for fee-bumping.
     /// PSBT Creator and Updater.
     pub fn new(
-        vault_input: VaultTxIn<VaultTxOut>,
-        feebump_input: Option<FeeBumpTxIn<FeeBumpTxOut>>,
+        vault_input: VaultTxIn,
+        feebump_input: Option<FeeBumpTxIn>,
         emer_txout: EmergencyTxOut,
         lock_time: u32,
     ) -> EmergencyTransaction {
@@ -465,8 +465,8 @@ impl UnvaultEmergencyTransaction {
     /// Script. It may also spend an additional output for fee-bumping.
     /// PSBT Creator and Updater.
     pub fn new(
-        unvault_input: UnvaultTxIn<UnvaultTxOut>,
-        feebump_input: Option<FeeBumpTxIn<FeeBumpTxOut>>,
+        unvault_input: UnvaultTxIn,
+        feebump_input: Option<FeeBumpTxIn>,
         emer_txout: EmergencyTxOut,
         lock_time: u32,
     ) -> UnvaultEmergencyTransaction {
@@ -499,7 +499,7 @@ impl SpendTransaction {
     /// txouts (including, but not restricted to, change).
     /// PSBT Creator and Updater.
     pub fn new(
-        unvault_inputs: Vec<UnvaultTxIn<UnvaultTxOut>>,
+        unvault_inputs: Vec<UnvaultTxIn>,
         spend_txouts: Vec<SpendTxOut>,
         lock_time: u32,
     ) -> SpendTransaction {
diff --git a/src/txins.rs b/src/txins.rs
index 9119a56..6f2caee 100644
--- a/src/txins.rs
+++ b/src/txins.rs
@@ -2,7 +2,7 @@
 //! Wrappers around bitcoin's OutPoint and previous TxOut to statically check Revault transaction
 //! creation and ease PSBT management.
 
-use crate::txouts::{CpfpTxOut, FeeBumpTxOut, RevaultTxOut, UnvaultTxOut, VaultTxOut};
+use crate::txouts::{CpfpTxOut, FeeBumpTxOut, UnvaultTxOut, VaultTxOut};
 
 use bitcoin::{OutPoint, TxIn};
 
@@ -21,25 +21,25 @@ pub trait RevaultTxIn<T>: fmt::Debug {
 }
 
 macro_rules! implem_revault_txin {
-    ( $struct_name:ident, $doc_comment:meta ) => {
+    ( $struct_name:ident, $txout_struct_name:ident, $doc_comment:meta ) => {
         #[$doc_comment]
         #[derive(Debug, Clone)]
-        pub struct $struct_name<T: RevaultTxOut> {
+        pub struct $struct_name {
             outpoint: OutPoint,
-            prev_txout: T,
+            prev_txout: $txout_struct_name,
             sequence: u32,
         }
 
-        impl<T: RevaultTxOut> RevaultTxIn<T> for $struct_name<T> {
+        impl RevaultTxIn<$txout_struct_name> for $struct_name {
             fn outpoint(&self) -> OutPoint {
                 self.outpoint
             }
 
-            fn txout(&self) -> &T {
+            fn txout(&self) -> &$txout_struct_name {
                 &self.prev_txout
             }
 
-            fn get_txout(self) -> T {
+            fn get_txout(self) -> $txout_struct_name {
                 self.prev_txout
             }
 
@@ -56,12 +56,13 @@ macro_rules! implem_revault_txin {
 
 implem_revault_txin!(
     VaultTxIn,
+    VaultTxOut,
     doc = "A vault txo spent by the unvault transaction and the emergency transaction"
 );
-impl VaultTxIn<VaultTxOut> {
+impl VaultTxIn {
     // FIXME: maybe we should take a reference to a VaultTransaction here to create the outpoint
     /// Instanciate a TxIn referencing a vault txout
-    pub fn new(outpoint: OutPoint, prev_txout: VaultTxOut, sequence: u32) -> VaultTxIn<VaultTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: VaultTxOut, sequence: u32) -> VaultTxIn {
         VaultTxIn {
             outpoint,
             prev_txout,
@@ -72,15 +73,12 @@ impl VaultTxIn<VaultTxOut> {
 
 implem_revault_txin!(
     UnvaultTxIn,
+    UnvaultTxOut,
     doc="An unvault txo spent by the cancel transaction, an emergency transaction, and the spend transaction."
 );
-impl UnvaultTxIn<UnvaultTxOut> {
+impl UnvaultTxIn {
     /// Instanciate a TxIn referencing an unvault txout
-    pub fn new(
-        outpoint: OutPoint,
-        prev_txout: UnvaultTxOut,
-        sequence: u32,
-    ) -> UnvaultTxIn<UnvaultTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: UnvaultTxOut, sequence: u32) -> UnvaultTxIn {
         UnvaultTxIn {
             outpoint,
             prev_txout,
@@ -91,17 +89,14 @@ impl UnvaultTxIn<UnvaultTxOut> {
 
 implem_revault_txin!(
     FeeBumpTxIn,
+    FeeBumpTxOut,
     doc = "A wallet txo spent by a revaulting (cancel, emergency) transaction to bump the transaction feerate.\
            This output is often created by a first stage transaction, but may directly be a wallet\
            utxo."
 );
-impl FeeBumpTxIn<FeeBumpTxOut> {
+impl FeeBumpTxIn {
     /// Instanciate a TxIn referencing a feebump txout
-    pub fn new(
-        outpoint: OutPoint,
-        prev_txout: FeeBumpTxOut,
-        sequence: u32,
-    ) -> FeeBumpTxIn<FeeBumpTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: FeeBumpTxOut, sequence: u32) -> FeeBumpTxIn {
         FeeBumpTxIn {
             outpoint,
             prev_txout,
@@ -112,11 +107,12 @@ impl FeeBumpTxIn<FeeBumpTxOut> {
 
 implem_revault_txin!(
     CpfpTxIn,
+    CpfpTxOut,
     doc = "The unvault CPFP txo spent to accelerate the confirmation of the unvault transaction."
 );
-impl CpfpTxIn<CpfpTxOut> {
+impl CpfpTxIn {
     /// Instanciate a TxIn referencing a CPFP txout
-    pub fn new(outpoint: OutPoint, prev_txout: CpfpTxOut, sequence: u32) -> CpfpTxIn<CpfpTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: CpfpTxOut, sequence: u32) -> CpfpTxIn {
         CpfpTxIn {
             outpoint,
             prev_txout,
```

darosior replied:
Done in last commit, can rebase it cleanly when i have a minute.

darosior commented:
Rebased on master and to clean up the history. Since 931274e  breaks bisectability i wonder if i should squash it on 8af4279.

Also added a commit at the root as we broke master again...

edouardparis commented on src/transactions.rs:
```diff
@@ -21,35 +28,284 @@ pub const RBF_SEQUENCE: u32 = u32::MAX - 2;
+                    bitcoin::Address::p2wsh(witness_script, bitcoin::Network::Bitcoin)
+                        .script_pubkey();
+                if expected_script_pubkey
+                    != psbtin
```
weird indentation, Is it rustfmt ?

darosior replied:
Yeah..

JSwambo commented on src/txins.rs:
```diff
@@ -0,0 +1,121 @@
+    /// Get a reference to the txout this txin refers
+    fn txout(&self) -> &T;
+    /// Get the actual txout this txin refers
+    fn get_txout(self) -> T;
```
rename to `as_outpoint`, `as_txout()`, `into_txout()` as per https://github.com/re-vault/revault/issues/19 ?

darosior replied:
I considered doing an actual cleanup PR for this and the documentation (maybe some more test cov as well). I could avoid introducing more here but... That incurs some non-trivial rebases (opened the issue after finishing this, fwiw)

JSwambo commented on src/txouts.rs:
```diff
@@ -12,28 +15,36 @@ pub trait RevaultTxOut: fmt::Debug + Clone + PartialEq {
     fn inner_txout(&self) -> &TxOut;
     /// Get the actual inner txout
     fn get_txout(self) -> TxOut;
```
Here too (https://github.com/re-vault/revault/issues/19)...  as_ and into_

JSwambo commented on src/transactions.rs:
```diff
@@ -21,35 +28,284 @@ pub const RBF_SEQUENCE: u32 = u32::MAX - 2;
+        Ok(())
+    }
+
+    /// Verify an input of the transaction against libbitcoinconsensus out of the informations
```
nit: information*

JSwambo reviewed (approved):
Ack 5e4dd40989c8ed42fac99ed8b572b6d5602bbb0b

edouardparis commented on src/transactions.rs:
```diff
@@ -21,35 +28,284 @@ pub const RBF_SEQUENCE: u32 = u32::MAX - 2;
+
+            // -- If a sighash type is provided, the signer must check that the sighash is acceptable.
+            // If unacceptable, they must fail.
+            if let Some(sighash_type) = rawsig.last().copied() {
```
I may be wrong, but what if rawsig has no sighash_type and last() returns last byte of the sig ?
Is .last() the best way to retrieve if exists, the sighash_type

edouardparis replied:
I may be confused by the comment `If a sighash type is provided..`

darosior replied:
Yeah, this assumes valid signature. This should probably not. At least i can type the rawsig parameter to take a secp256k1 sig and bitcoin::SighashType tuple.

But should we verify the sig here ? Hmmm.
My main reason for not adding it here was that i don't want us to rely on this contextless verification downstream. OTOH we don't get the sighash from our peer, instead we create it so we may have enough context here..

Sorry for the useless comment, thinking out loud here!

darosior replied:
@JSwambo wdyt ?

darosior replied:
> I may be confused by the comment If a sighash type is provided..

Copied it straight from the BIP, which advises to check signatures here and which i purposefully did not do for the aforementioned reason. Now i'm wondering if checking the sighash but not the sig makes sense..

darosior replied:
Went for the `BitcoinSig` type but no sig check, see fixup commit.

edouardparis commented:
ACK 30305bcc3e7fdf93d4610248e8c03bb839883827

edouardparis reviewed (approved):
ACK 30305bcc3e7fdf93d4610248e8c03bb839883827

darosior commented:
Ok so:
- improved the API for the `add_signature()` method as pointed by @edouardparis 
- addressed @JSwambo nits
- squashed some commits to maintain bisectability (`rebase master -x "cargo test"`)

Can you re review please ?

edouardparis reviewed (approved):
ACK a0e0ff54f655945116f5515907bb4215197671dd

darosior commented:
Ok, missed the sequence stuff. Merging this and doing a follow-up.
